### PR TITLE
Implementation of a network element impact and a modification result

### PR DESCRIPTION
**Pull request by SlimaneAmar**

### CODE REVIEW

File: src/main/java/org/gridsuite/study/server/notification/dto/NetworkImpcatsInfos.java [hunk 1]
```diff
@@ -9,19 +9,19 @@
...
-import lombok.Setter;
+import lombok.*;
 
 @AllArgsConstructor
 @Setter
 @Getter
 @Builder
 @NoArgsConstructor
+@ToString
 public class NetworkImpcatsInfos {
```
[the remark this answers is not in the document]
Comment (SlimaneAmar) [reply src/main/java/org/gridsuite/study/server/notification/dto/NetworkImpcatsInfos.java]: Done

File: src/main/java/org/gridsuite/study/server/service/StudyService.java [hunk 2]
```diff
@@ -1473,51 +1471,32 @@ public void deleteNodeReport(UUID nodeUuid) {
...
+        notificationService.emitStudyChanged(studyUuid, nodeUuid, NotificationService.UPDATE_TYPE_STUDY,
+            NetworkImpcatsInfos.builder()
+                .deletedEquipments(deletionsInfos)
+                .impactedSubstationsIds(networkModificationResult.getImpactedSubstationsIds())
+                .build()
+        );
+
+        updateStatuses(studyUuid, nodeUuid);
+
+        if (networkModificationResult.getNetworkImpacts().stream()
```
[the remark this answers is not in the document]
Comment (SlimaneAmar) [reply src/main/java/org/gridsuite/study/server/service/StudyService.java:1491]: Yes

File: src/test/java/org/gridsuite/study/server/utils/ImpactUtils.java [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+ * @author Slimane Amar <slimane.amar at rte-france.com>
+ */
+public final class ImpactUtils {
+
+    private ImpactUtils() {
+    }
+
+    public static Optional<NetworkModificationResult> createModificationResultWithElementImpact(SimpleImpactType impactType, IdentifiableType elementType, String elementId, Set<String> substationIds) {
+        return Optional.of(NetworkModificationResult.builder()
+            .networkImpacts(List.of(createElementImpact(impactType, elementType, elementId, substationIds)))
```
Comment (flomillot) [inline src/test/java/org/gridsuite/study/server/utils/ImpactUtils.java:28]: You never set the application status ?

File: src/test/java/org/gridsuite/study/server/StudyTest.java [hunk 4]
```diff
@@ -1229,23 +1230,26 @@ public void testDuplicateStudy() throws Exception {
...
+            createModificationResultWithElementImpact(SimpleImpactType.CREATION, IdentifiableType.TWO_WINDINGS_TRANSFORMER, "2wtId", Set.of("s1"));
+        UUID stubPostId = wireMockUtils.stubNetworkModificationPost(mapper.writeValueAsString(networkModificationResult));
         mockMvc.perform(post(URI_NETWORK_MODIF, study1Uuid, node1.getId())
                 .contentType(MediaType.APPLICATION_JSON)
                 .content(createTwoWindingsTransformerAttributes)
                 .header(USER_ID_HEADER, "userId"))
                 .andExpect(status().isOk());
-        NetworkImpcatsInfos expectedPayload = NetworkImpcatsInfos.builder().impactedSubstationsIds(ImmutableSet.of("s2")).deletedEquipments(ImmutableSet.of()).build();
+        NetworkImpactsInfos expectedPayload = NetworkImpactsInfos.builder().impactedSubstationsIds(ImmutableSet.of("s1")).deletedEquipments(ImmutableSet.of()).build();
         checkEquipmentCreatingMessagesReceived(study1Uuid, node1.getId());
```
Comment (flomillot) [inline src/test/java/org/gridsuite/study/server/StudyTest.java:1242]: It's not related to this PR but why we have 2 "checkEquipmentCreatingMessagesReceived", we can't mix it into one method ?

File: src/test/java/org/gridsuite/study/server/utils/ImpactUtils.java (hunk 3 above)
Comment (flomillot) [reply src/test/java/org/gridsuite/study/server/utils/ImpactUtils.java:28]: I'm doing it in my PR about build status

File: src/test/java/org/gridsuite/study/server/StudyTest.java [hunk 5]
```diff
@@ -1340,23 +1335,29 @@ public void testCutAndPasteNode() throws Exception {
 
         // add modification on node "node1"
         String createTwoWindingsTransformerAttributes = "{\"type\":\"" + ModificationType.TWO_WINDINGS_TRANSFORMER_CREATION + "\",\"equipmentId\":\"2wtId\",\"equipmentName\":\"2wtName\",\"seriesResistance\":\"10\",\"seriesReactance\":\"10\",\"magnetizingConductance\":\"100\",\"magnetizingSusceptance\":\"100\",\"ratedVoltage1\":\"480\",\"ratedVoltage2\":\"380\",\"voltageLevelId1\":\"CHOO5P6\",\"busOrBusbarSectionId1\":\"CHOO5P6_1\",\"voltageLevelId2\":\"CHOO5P6\",\"busOrBusbarSectionId2\":\"CHOO5P6_1\"}";
+        String networkModificationResultJson = mapper.writeValueAsString(Optional.of(NetworkModificationResult.builder()
```
Comment (flomillot) [inline src/test/java/org/gridsuite/study/server/StudyTest.java:1338]: The node is not built so you shouldn't return a NetworkModificationResult

File: src/main/java/org/gridsuite/study/server/dto/modification/NetworkModificationResult.java [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+    @Schema(description = "Application status")
+    ApplicationStatus applicationStatus;
+
+    @Schema(description = "Network modification impacts")
+    @Builder.Default
+    private List<SimpleElementImpact> networkImpacts = List.of();
+
+    public Set<String> getImpactedSubstationsIds() {
+        return networkImpacts.stream()
+            .filter(impact -> impact.getImpactType() != SimpleElementImpact.SimpleImpactType.DELETION || impact.getElementType() != IdentifiableType.SUBSTATION)
```
Comment (ayolab) [inline src/main/java/org/gridsuite/study/server/dto/modification/NetworkModificationResult.java:44]:
```suggestion
            .filter(impact -> impact.getImpactType() != SimpleElementImpact.SimpleImpactType.DELETION && impact.getElementType() != IdentifiableType.SUBSTATION)
```

**Review by ayolab — CHANGES_REQUESTED** [review]

File: src/main/java/org/gridsuite/study/server/service/StudyService.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (SlimaneAmar) [reply src/main/java/org/gridsuite/study/server/service/StudyService.java:1491]: No it's not possible a java stream can only be consumed once

File: src/test/java/org/gridsuite/study/server/Impacts/ElementImpactTest.java [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+    public void testNetworkModificationResult() {
+        EqualsVerifier.simple().forClass(SimpleElementImpact.class).verify();
+
+        SimpleElementImpact creationImpact = createElementImpact(SimpleImpactType.CREATION, IdentifiableType.LINE, "lineId", new TreeSet<>(List.of("s2", "s1")));
+        SimpleElementImpact modificationImpact = createElementImpact(SimpleImpactType.MODIFICATION, IdentifiableType.LOAD, "loadId", new TreeSet<>(List.of("s3")));
+        SimpleElementImpact deletionImpact = createElementImpact(SimpleImpactType.DELETION, IdentifiableType.GENERATOR, "generatorId", new TreeSet<>(List.of("s4")));
+
+        Collection<SimpleElementImpact> impacts = List.of(creationImpact, modificationImpact, deletionImpact);
+
+        assertEquals("{\"impactType\":\"CREATION\",\"elementId\":\"lineId\",\"elementType\":\"LINE\",\"substationIds\":[\"s1\",\"s2\"]}", mapper.writeValueAsString(creationImpact));
```
[the remark this answers is not in the document]
Comment (SlimaneAmar) [reply src/test/java/org/gridsuite/study/server/Impacts/ElementImpactTest.java:44]: It is to test the json of a dto in respect of the annotations
For the moment there are no particular annotations but maybe later

File: src/test/java/org/gridsuite/study/server/Impacts/ElementImpactTest.java [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+        assertEquals("{\"impactType\":\"MODIFICATION\",\"elementId\":\"loadId\",\"elementType\":\"LOAD\",\"substationIds\":[\"s3\"]}", mapper.writeValueAsString(modificationImpact));
+        assertEquals("{\"impactType\":\"DELETION\",\"elementId\":\"generatorId\",\"elementType\":\"GENERATOR\",\"substationIds\":[\"s4\"]}", mapper.writeValueAsString(deletionImpact));
+
+        NetworkModificationResult result = NetworkModificationResult.builder()
+            .applicationStatus(NetworkModificationResult.ApplicationStatus.ALL_OK)
+            .networkImpacts((List<SimpleElementImpact>) impacts)
+            .build();
+        assertEquals(
+            "{\"applicationStatus\":\"ALL_OK\",\"networkImpacts\":[{\"impactType\":\"CREATION\",\"elementId\":\"lineId\",\"elementType\":\"LINE\",\"substationIds\":[\"s1\",\"s2\"]},{\"impactType\":\"MODIFICATION\",\"elementId\":\"loadId\",\"elementType\":\"LOAD\",\"substationIds\":[\"s3\"]},{\"impactType\":\"DELETION\",\"elementId\":\"generatorId\",\"elementType\":\"GENERATOR\",\"substationIds\":[\"s4\"]}],\"impactedSubstationsIds\":[\"s1\",\"s2\",\"s3\",\"s4\"]}",
```
[the remark this answers is not in the document]
Comment (SlimaneAmar) [reply src/test/java/org/gridsuite/study/server/Impacts/ElementImpactTest.java:53]: Done

File: src/test/java/org/gridsuite/study/server/NetworkModificationTest.java [hunk 9]
```diff
@@ -1789,191 +1722,100 @@ public void testDeleteVoltageLevelOnline() {
...
                         .andExpect(status().isBadRequest());
         checkEquipmentUpdatingMessagesReceived(studyNameUserIdUuid, modificationNodeUuid);
         checkEquipmentUpdatingFinishedMessagesReceived(studyNameUserIdUuid, modificationNodeUuid);
-        verifyNetworkModificationPut(stubIdPutErr, badBody);
-    }
-
-    @SneakyThrows
-    @Test
-    @DisplayName("Should throw an exception when the body of a network modification is wrongly formatted")
-    public void testNetworkModificationBodyBadFormat() {
```
[the remark this answers is not in the document]
Comment (SlimaneAmar) [reply src/test/java/org/gridsuite/study/server/NetworkModificationTest.java:1896]: The method `getModificationType` has been removed

File: src/test/java/org/gridsuite/study/server/utils/ImpactUtils.java (hunk 3 above)
Comment (SlimaneAmar) [reply src/test/java/org/gridsuite/study/server/utils/ImpactUtils.java:28]: Done

File: src/test/java/org/gridsuite/study/server/StudyTest.java (hunk 4 above)
Comment (SlimaneAmar) [reply src/test/java/org/gridsuite/study/server/StudyTest.java:1234]: checkEquipment**Creation**MessagesReceived ?

File: src/main/java/org/gridsuite/study/server/dto/modification/NetworkModificationResult.java (hunk 6 above)
Comment (SlimaneAmar) [reply src/main/java/org/gridsuite/study/server/dto/modification/NetworkModificationResult.java:44]: Done

File: src/test/java/org/gridsuite/study/server/StudyTest.java (hunk 5 above)
Comment (SlimaneAmar) [reply src/test/java/org/gridsuite/study/server/StudyTest.java:1338]: Done

**Review by ayolab — APPROVED** [review]
Tests OK
Code OK

File: src/main/java/org/gridsuite/study/server/service/StudyService.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (flomillot) [reply src/main/java/org/gridsuite/study/server/service/StudyService.java:1491]: Ah ok my bad !

File: src/test/java/org/gridsuite/study/server/NetworkModificationTest.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (flomillot) [reply src/test/java/org/gridsuite/study/server/NetworkModificationTest.java:1896]: Then we need to adapt code / tests, but not removing the tests

File: src/test/java/org/gridsuite/study/server/StudyTest.java (hunk 4 above)
Comment (flomillot) [reply src/test/java/org/gridsuite/study/server/StudyTest.java:1234]: Ah I misread, my bad.

**Review by flomillot — CHANGES_REQUESTED** [review]
Waiting for renaming or Javadoc of SimpleElementImpact
edit : what about the 2 removed tests ?

**Review by flomillot — APPROVED** [review]

**Pull request merged**